onnxruntime/core/optimizer/matmul_add_fusion.cc

```diff
@@ -43,10 +43,13 @@ Status MatMulAddFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level)
     auto matmul_input_defs = matmul_node.MutableInputDefs();
     auto add_input_defs = add_node.MutableInputDefs();
 
-    // Gemm only support float, so the inputs of MatMul
+    // Gemm requires that inputs be the same data type and both floating point (float32/float16).
     auto matmul_type = matmul_input_defs[0]->Type();
     auto add_type = add_input_defs[0]->Type();
-    if ((*matmul_type) != "tensor(float)" || (*add_type) != "tensor(float)") {
+    if ((*matmul_type) != (*add_type)) {
+      continue;
+    }
+    if ((*matmul_type) != "tensor(float)" && (*matmul_type) != "tensor(float16)") {
```

Comment:
Can double be added either?

Reply:
@weixingzhang : Possibly. Would need to test float64 GEMM...

Reply:
@weixingzhang : Trying float64, I get:
`HRESULT=0x80004001 message=Could not find an implementation for the node MatMul(7)`
`HRESULT=0x80004001 message=Could not find an implementation for the node Add(7)`
And if I try GEMM directly (no fusion), I get:
`Error: HRESULT=0x80004001 message=Could not find an implementation for the node Gemm(7)`

So it seems no kernel is registered for float64 (neither opset 7 or opset 9); and the cast nodes evidently do not convert between float64 either. I can still add float64 too in the graph transformer, but it makes no difference either way because MatMul.